radix-engine-interface/src/model/resource/non_fungible_local_id.rs

```diff
@@ -14,78 +14,266 @@ use utils::copy_u8_array;
-///
-/// In the future, we will use dedicated types to represent each variant.
 pub trait IsNonAutoGeneratedNonFungibleLocalId: IsNonFungibleLocalId {}
 
-impl IsNonAutoGeneratedNonFungibleLocalId for String {}
-impl IsNonAutoGeneratedNonFungibleLocalId for u64 {}
-impl IsNonAutoGeneratedNonFungibleLocalId for Vec<u8> {}
+impl IsNonAutoGeneratedNonFungibleLocalId for StringNonFungibleLocalId {}
+impl IsNonAutoGeneratedNonFungibleLocalId for IntegerNonFungibleLocalId {}
+impl IsNonAutoGeneratedNonFungibleLocalId for BytesNonFungibleLocalId {}
 
-impl From<String> for NonFungibleLocalId {
-    fn from(value: String) -> Self {
-        Self::String(value)
+impl TryFrom<String> for NonFungibleLocalId {
+    type Error = ContentValidationError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Ok(StringNonFungibleLocalId::new(value)?.into())
     }
 }
+
 impl From<u64> for NonFungibleLocalId {
     fn from(value: u64) -> Self {
-        Self::Integer(value)
+        IntegerNonFungibleLocalId::new(value).into()
     }
 }
-impl From<Vec<u8>> for NonFungibleLocalId {
-    fn from(value: Vec<u8>) -> Self {
-        Self::Bytes(value)
+
+impl TryFrom<Vec<u8>> for NonFungibleLocalId {
+    type Error = ContentValidationError;
+
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        Ok(BytesNonFungibleLocalId::new(value)?.into())
     }
 }
-impl From<u128> for NonFungibleLocalId {
-    fn from(value: u128) -> Self {
-        Self::UUID(value)
+
+impl TryFrom<u128> for NonFungibleLocalId {
+    type Error = ContentValidationError;
+
+    fn try_from(value: u128) -> Result<Self, Self::Error> {
+        Ok(UUIDNonFungibleLocalId::new(value)?.into())
     }
 }
 
 /// Represents the local id of a non-fungible.
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub enum NonFungibleLocalId {
-    /// String of `[_0-9a-zA-Z]{1,64}`.
-    String(String),
+    /// String matching `[_0-9a-zA-Z]{1,64}`.
+    ///
+    /// Create using `NonFungibleLocalId::string(...).unwrap()`.
+    String(StringNonFungibleLocalId),
     /// Unsigned integers, up to u64.
-    Integer(u64),
+    ///
+    /// Create using `NonFungibleLocalId::integer(...)`.
+    Integer(IntegerNonFungibleLocalId),
     /// Bytes, of length between 1 and 64.
-    Bytes(Vec<u8>),
+    ///
+    /// Create using `NonFungibleLocalId::bytes(...).unwrap()`.
+    Bytes(BytesNonFungibleLocalId),
     /// UUID, v4, variant 1, big endian. See https://www.rfc-editor.org/rfc/rfc4122
-    UUID(u128),
+    ///
+    /// Create using `NonFungibleLocalId::uuid(...).unwrap()`.
+    UUID(UUIDNonFungibleLocalId),
+}
+
+impl NonFungibleLocalId {
+    pub fn string<T: Into<String>>(value: T) -> Result<Self, ContentValidationError> {
+        value.into().try_into()
+    }
+
+    pub fn integer(value: u64) -> Self {
+        value.into()
+    }
+
+    pub fn bytes<T: Into<Vec<u8>>>(value: T) -> Result<Self, ContentValidationError> {
```

Comment:
Hmm I think that makes sense for methods which need to take a `&[u8]` slice, but less so for methods which need to take an owned Vec - otherwise passing an owned Vec into a `AsRef<[u8]>` would require it be cloned.

Are you happy to keep it as-is? :)

Reply:
Cool!